File: common/src/main/java/weechan/com/common/utils/loading/Loading.java
package weechan.com.common.utils.loading;

import android.graphics.Rect;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.IdRes;
import weechan.com.common.R;

import static android.view.View.INVISIBLE;
import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static weechan.com.common.utils.loading.LoadingTag.ERROR_STATE;
import static weechan.com.common.utils.loading.LoadingTag.LOADING_STATE;
import static weechan.com.common.utils.loading.LoadingTag.NONE_STATE;
import static weechan.com.common.utils.loading.LoadingTag.RETRY_STATE;
import static weechan.com.common.utils.loading.LoadingTag.WRAPPER_TAG;


/**
 * @author 214652773@qq.com
 * @user c
 * @create 2018/11/9 10:45
 */

public class Loading {

    private static final int MAGIC_NUMBER = R.id.loading_tag_state;

    private LoadingBuilder mBuilder;

    private View target;
//    private static Loading mLoading = new Loading();

    public Loading(LoadingBuilder loadingBuilder) {
        mBuilder = loadingBuilder;
        target = mBuilder.getTarget();
    }

    public void load() {
        display(target, getCurrentSate(target), LOADING_STATE);
    }

    public void error() {
        display(target, getCurrentSate(target), ERROR_STATE);
    }

    public void retry() {
        display(target, getCurrentSate(target), RETRY_STATE);
    }

    public void finish() {
        display(target, getCurrentSate(target), NONE_STATE);
    }


    public <T> T findViewById(@IdRes int id){
        LoadingWrapper wrapper = ((ViewGroup) target.getParent())
                .findViewWithTag(WRAPPER_TAG + target.hashCode());
        return (T)wrapper.findViewById(id);
    }

    private int getCurrentSate(View target) {
        Integer state = (Integer) target.getTag(MAGIC_NUMBER);
        if (state == null) {
            state = NONE_STATE;
        }
        return state;
    }

    private void display(final View target, final int currentState, int targetState) {
        Log.e("STATE", currentState + "  " + targetState);
        if (targetState == currentState) return;

        LoadingWrapper wrapper = ((ViewGroup) target.getParent())
                .findViewWithTag(WRAPPER_TAG + target.hashCode());

        if (targetState == NONE_STATE) {
            wrapper.activeViewByState(targetState);
            wrapper.setVisibility(View.GONE);
            target.setVisibility(View.VISIBLE);
            target.setTag(MAGIC_NUMBER, NONE_STATE);
            return;
        }

        //避免移动的
        target.setVisibility(INVISIBLE);

        //wrapper不存在需要创建
        if (wrapper == null) {
            final Rect border = new Rect(target.getLeft(), target.getTop(),
                    target.getRight(), target.getBottom());

            wrapper = new LoadingWrapper(target.getContext(), border,
                    mBuilder.getLoadingView(),
                    mBuilder.getErrorView(),
                    mBuilder.getRetryView());
           // wrapper.setLayoutParams(new ViewGroup.LayoutParams(0,0));

            wrapper.setTag(WRAPPER_TAG + target.hashCode());

            wrapper.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mBuilder == null) return;
                    switch (getCurrentSate(target)) {
                        case LOADING_STATE:
                            if (mBuilder.getOnLoadingClick() != null)
                                mBuilder.getOnLoadingClick().run();
                            break;
                        case LoadingTag.ERROR_STATE:
                            if (mBuilder.getOnErrorClick() != null)
                                mBuilder.getOnErrorClick().run();
                            break;
                        case LoadingTag.RETRY_STATE:
                            if (mBuilder.getOnRetryClick() != null)
                                mBuilder.getOnRetryClick().run();
                            break;
                    }
                }
            });

            ((ViewGroup) target.getParent()).addView(wrapper);
        }

        wrapper.setVisibility(INVISIBLE);

        wrapper.activeViewByState(targetState);

        //改变当前状态为目标状态
        target.setTag(MAGIC_NUMBER, targetState);

        final LoadingWrapper finalWrapper = wrapper;
        wrapper.post(new Runnable() {
            @Override
            public void run() {
                finalWrapper.setVisibility(View.VISIBLE);
                ViewGroup.LayoutParams params = finalWrapper.getLayoutParams();

//                finalWrapper.setLayoutParams(params);
            }
        });

    }
}

File: app/src/main/java/weechan/com/whatsforlunch/widget/Loading.kt
//package weechan.com.whatsforlunch.widget
//
//import android.content.Context
//import android.util.Log
//import android.view.Gravity
//import android.view.LayoutInflater
//import android.view.View
//import android.view.View.*
//import android.view.ViewGroup
//import android.view.ViewGroup.LayoutParams.MATCH_PARENT
//import android.widget.FrameLayout
//import androidx.annotation.LayoutRes
//import weechan.com.whatsforlunch.R
//import weechan.com.whatsforlunch.widget.LoadingTag.*
//
///**
// *
// * @author 214652773@qq.com
// * @user c
// * @create 2018/11/7 12:58
// *
// */
//
//
//class Loading(private val loadingView : View) {
//
//    companion object {
//        private val MAGIC_NUMBER = R.id.loading_tag_state
//        private val loading = Loading()
//
//        fun load(target: View) = loading.load(target)
//        fun error(target: View) = loading.load(target)
//        fun retry(target: View) = loading.load(target)
//    }
//
//
//    fun load(@LayoutRes layoutId: Int, context: Context) {
//        val view = LayoutInflater.from(context).inflate(layoutId, null, false)
//    }
//
//    fun load(target: View) {
//        display(target, getCurrentSate(target), LOADING_STATE)
//    }
//
//    fun error(target: View) {
//        display(target, getCurrentSate(target), ERROR_STATE)
//    }
//
//    fun retry(target: View) {
//        display(target, getCurrentSate(target), ERROR_STATE)
//    }
//
//    private fun getCurrentSate(target: View): String {
//        var state = target.getTag(MAGIC_NUMBER) as String?
//        if (state == null) {
//            state = NONE_STATE
//        }
//        return state
//    }
//
//    private fun display(target: View, currentState: String, targetState: String) {
//        if (targetState == currentState) return
//
//        target.visibility = INVISIBLE
//        var wrapper = (target.parent as ViewGroup)
//                .findViewWithTag<FrameLayout>(WRAPPER_TAG + target.hashCode())
//
//        //wrapper不存在需要创建
//        if (wrapper == null) {
//            wrapper = LoadingWrapper(target.context, target.left, target.top, target.right, target.bottom).apply {
//                tag = WRAPPER_TAG + target.hashCode()
//            }
//            (target.parent as ViewGroup).addView(wrapper)
//        }
//
//        wrapper.visibility = INVISIBLE
//
//        with(wrapper) {
//            val viewToShow = findViewWithTag<View>(targetState)
//            val currentView = findViewWithTag<View>(currentState)
//
//            currentView?.visibility = GONE
//            viewToShow.visibility = VISIBLE
//
//            //改变当前状态为目标状态
//            target.setTag(MAGIC_NUMBER, targetState)
//        }
//    }
//
//
//
//}
//
////class loading{
////
////        fun load(view: ViewGroup, loadingView: View) {
////            val viewWithTag = getInsertedView(view).findViewWithTag<View>("com.weechan.loading.loadTag")
////            if (viewWithTag != null) return
////
////            checkType(loadingView)
////
////            loadingView.apply { setTag("com.weechan.loading.loadTag") }
////            dismiss(view)
////            if (view is androidx.recyclerview.widget.RecyclerView || view is ListView) {
////                loadList(view.parent as ViewGroup, loadingView, view)
////            } else {
////                loadNormal(view, loadingView)
////            }
////        }
////
////        fun error(view: ViewGroup, errorView: View) {
////
////            val viewWithTag = getInsertedView(view).findViewWithTag<View>("com.weechan.loading.errorTag")
////            if (viewWithTag != null) return
////
////            checkType(errorView)
////
////            errorView.setTag("com.weechan.loading.errorTag")
////            dismiss(view)
////            if (view is androidx.recyclerview.widget.RecyclerView || view is ListView) {
////                loadList(view.parent as ViewGroup, errorView, view)
////            } else {
////                loadNormal(view, errorView)
////            }
////        }
////
////        fun finish(view: ViewGroup) {
////            dismiss(view)
////            if (view is androidx.recyclerview.widget.RecyclerView || view is ListView) {
////                view.visibility = View.VISIBLE
////            } else {
////                val container = view.parent as ViewGroup
////                container.forEachChild {
////                    it.visibility = View.VISIBLE
////                }
////            }
////        }
////
////        private fun checkType(view: View) {
////            assert(view is RelativeView)
////        }
////
////        private fun dismiss(view: ViewGroup) {
////            val container = getInsertedView(view)
////            arrayOf("com.weechan.loading.loadTag", "com.weechan.loading.errorTag").forEach {
////                container.findViewWithTag<View>(it)?.let { existView ->
////                    container.removeView(existView)
////                }
////            }
////        }
////
////        private fun getInsertedView(view: ViewGroup): ViewGroup {
////            if (view is androidx.recyclerview.widget.RecyclerView || view is ListView) {
////                return view.parent as ViewGroup
////            }
////            return view;
////        }
////
////        private fun loadList(parent: ViewGroup, loadingView: View, listView: ViewGroup) {
////            listView.visibility = View.INVISIBLE
////            val left = listView.left
////            val top = listView.top
////            loadingView.layoutParams = ViewGroup.MarginLayoutParams(WRAP_CONTENT, WRAP_CONTENT)
////            loadingView.measure(0, 0)
////            val width = loadingView.measuredWidth;
////            val height = loadingView.measuredHeight
////
////            val loadingViewLeft = left + (listView.width - width) / 2
////            val loadingViewTop = top + (listView.height - height) / 2
////
////            parent.addView(loadingView)
////            parent.post {
////                (loadingView as RelativeView).absLayout(loadingViewLeft, loadingViewTop,
////                        loadingViewLeft + width, loadingViewTop + height)
////            }
////        }
////
////        private fun loadNormal(parent: ViewGroup, loadingView: View) {
////
////            for (index in 0 until parent.childCount) {
////                parent.getChildAt(index).visibility = View.INVISIBLE
////            }
////
////            val params = ViewGroup.MarginLayoutParams(WRAP_CONTENT, WRAP_CONTENT)
////            val (left, top) = arrayOf((parent.width - loadingView.measuredWidth) / 2,
////                    (parent.height - loadingView.measuredHeight) / 2)
////            loadingView.measure(0, 0)
////            loadingView.layoutParams = params
////            parent.addView(loadingView)
////
////            val (halfWidth, halfHeight) = arrayOf(loadingView.measuredWidth / 2, loadingView.measuredHeight / 2)
////            loadingView.post {
////                (loadingView as RelativeView).absLayout(left - halfWidth, top - halfHeight,
////                        left + halfWidth, top + halfHeight)
////            }
////        }
//
////}